File: ios/Reguerta/Reguerta/Domain/Access/MemberPermissionMatrix.swift
import Foundation

enum AccessCapability: String, CaseIterable, Sendable {
    case accessCommonHomeModules = "access_common_home_modules"
    case manageProductCatalog = "manage_product_catalog"
    case accessReceivedOrders = "access_received_orders"
    case manageMembers = "manage_members"
    case grantAdminRole = "grant_admin_role"
    case publishNews = "publish_news"
    case sendAdminNotifications = "send_admin_notifications"
    case routeProductionReviewerToDevelop = "route_production_reviewer_to_develop"
}

enum CanonicalAccessRole: String, CaseIterable, Sendable {
    case member
    case producer
    case admin
    case reviewer
}

enum MemberPermissionMatrix {
    private static let roleCapabilities: [CanonicalAccessRole: Set<AccessCapability>] = [
        .member: [.accessCommonHomeModules],
        .producer: [.accessCommonHomeModules, .manageProductCatalog, .accessReceivedOrders],
        .admin: [.accessCommonHomeModules, .manageMembers, .grantAdminRole, .publishNews, .sendAdminNotifications],
        .reviewer: [.accessCommonHomeModules, .routeProductionReviewerToDevelop],
    ]

    static var reviewerCapabilities: Set<AccessCapability> {
        capabilities(for: .reviewer)
    }

    static func capabilities(for role: CanonicalAccessRole) -> Set<AccessCapability> {
        roleCapabilities[role, default: []]
    }

    static func capabilities(for role: MemberRole) -> Set<AccessCapability> {
        switch role {
        case .member:
            capabilities(for: CanonicalAccessRole.member)
        case .producer:
            capabilities(for: CanonicalAccessRole.producer)
        case .admin:
            capabilities(for: CanonicalAccessRole.admin)
        }
    }

    static func capabilities(for member: Member) -> Set<AccessCapability> {
        var merged = member.roles.reduce(into: Set<AccessCapability>()) { partialResult, role in
            partialResult.formUnion(capabilities(for: role))
        }
        if member.isCommonPurchaseManager {
            merged.insert(.manageProductCatalog)
        }
        return merged
    }

    static func hasCapability(_ capability: AccessCapability, for member: Member) -> Bool {
        capabilities(for: member).contains(capability)
    }
}

extension Member {
    var isMember: Bool {
        roles.contains(.member)
    }

    var isProducer: Bool {
        roles.contains(.producer)
    }

    var canAccessCommonHomeModules: Bool {
        MemberPermissionMatrix.hasCapability(.accessCommonHomeModules, for: self)
    }

    var canManageProductCatalog: Bool {
        MemberPermissionMatrix.hasCapability(.manageProductCatalog, for: self)
    }

    var canAccessReceivedOrders: Bool {
        MemberPermissionMatrix.hasCapability(.accessReceivedOrders, for: self)
    }

    var canManageMembers: Bool {
        MemberPermissionMatrix.hasCapability(.manageMembers, for: self)
    }

    var canGrantAdminRole: Bool {
        MemberPermissionMatrix.hasCapability(.grantAdminRole, for: self)
    }

    var canPublishNews: Bool {
        MemberPermissionMatrix.hasCapability(.publishNews, for: self)
    }

    var canSendAdminNotifications: Bool {
        MemberPermissionMatrix.hasCapability(.sendAdminNotifications, for: self)
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/access/MemberPermissionMatrix.kt
package com.reguerta.user.domain.access

enum class AccessCapability(val wireValue: String) {
    ACCESS_COMMON_HOME_MODULES("access_common_home_modules"),
    MANAGE_PRODUCT_CATALOG("manage_product_catalog"),
    ACCESS_RECEIVED_ORDERS("access_received_orders"),
    MANAGE_MEMBERS("manage_members"),
    GRANT_ADMIN_ROLE("grant_admin_role"),
    PUBLISH_NEWS("publish_news"),
    SEND_ADMIN_NOTIFICATIONS("send_admin_notifications"),
    ROUTE_PRODUCTION_REVIEWER_TO_DEVELOP("route_production_reviewer_to_develop"),
    ;

    companion object {
        private val byWireValue = entries.associateBy(AccessCapability::wireValue)

        fun fromWireValue(value: String): AccessCapability? = byWireValue[value]
    }
}

enum class CanonicalAccessRole(val wireValue: String) {
    MEMBER("member"),
    PRODUCER("producer"),
    ADMIN("admin"),
    REVIEWER("reviewer"),
    ;

    companion object {
        private val byWireValue = entries.associateBy(CanonicalAccessRole::wireValue)

        fun fromWireValue(value: String): CanonicalAccessRole? = byWireValue[value]
    }
}

object MemberPermissionMatrix {
    private val roleCapabilities = mapOf(
        CanonicalAccessRole.MEMBER to setOf(
            AccessCapability.ACCESS_COMMON_HOME_MODULES,
        ),
        CanonicalAccessRole.PRODUCER to setOf(
            AccessCapability.ACCESS_COMMON_HOME_MODULES,
            AccessCapability.MANAGE_PRODUCT_CATALOG,
            AccessCapability.ACCESS_RECEIVED_ORDERS,
        ),
        CanonicalAccessRole.ADMIN to setOf(
            AccessCapability.ACCESS_COMMON_HOME_MODULES,
            AccessCapability.MANAGE_MEMBERS,
            AccessCapability.GRANT_ADMIN_ROLE,
            AccessCapability.PUBLISH_NEWS,
            AccessCapability.SEND_ADMIN_NOTIFICATIONS,
        ),
        CanonicalAccessRole.REVIEWER to setOf(
            AccessCapability.ACCESS_COMMON_HOME_MODULES,
            AccessCapability.ROUTE_PRODUCTION_REVIEWER_TO_DEVELOP,
        ),
    )

    fun capabilitiesFor(role: CanonicalAccessRole): Set<AccessCapability> =
        roleCapabilities.getValue(role)

    fun capabilitiesFor(role: MemberRole): Set<AccessCapability> =
        when (role) {
            MemberRole.MEMBER -> capabilitiesFor(CanonicalAccessRole.MEMBER)
            MemberRole.PRODUCER -> capabilitiesFor(CanonicalAccessRole.PRODUCER)
            MemberRole.ADMIN -> capabilitiesFor(CanonicalAccessRole.ADMIN)
        }

    fun capabilitiesFor(member: Member): Set<AccessCapability> {
        val merged = member.roles
            .asSequence()
            .flatMap { role -> capabilitiesFor(role).asSequence() }
            .toMutableSet()
        if (member.isCommonPurchaseManager) {
            merged += AccessCapability.MANAGE_PRODUCT_CATALOG
        }
        return merged
    }

    fun hasCapability(member: Member, capability: AccessCapability): Boolean =
        capabilitiesFor(member).contains(capability)

    val reviewerCapabilities: Set<AccessCapability>
        get() = capabilitiesFor(CanonicalAccessRole.REVIEWER)
}

val Member.isMember: Boolean
    get() = roles.contains(MemberRole.MEMBER)

val Member.isProducer: Boolean
    get() = roles.contains(MemberRole.PRODUCER)

val Member.canAccessCommonHomeModules: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.ACCESS_COMMON_HOME_MODULES)

val Member.canManageProductCatalog: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.MANAGE_PRODUCT_CATALOG)

val Member.canAccessReceivedOrders: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.ACCESS_RECEIVED_ORDERS)

val Member.canManageMembers: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.MANAGE_MEMBERS)

val Member.canGrantAdminRole: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.GRANT_ADMIN_ROLE)

val Member.canPublishNews: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.PUBLISH_NEWS)

val Member.canSendAdminNotifications: Boolean
    get() = MemberPermissionMatrix.hasCapability(this, AccessCapability.SEND_ADMIN_NOTIFICATIONS)
